driver/src/io_backend/ext/xaie_amdair.c

```diff
@@ -0,0 +1,615 @@
+* @param	IOInst: IO instance pointer
+* @param	RegOff: Register offset to read from.
+* @param	Value: 32-bit data to be written.
+*
+* @return	None.
+*
+* @note		None.
+*
+*******************************************************************************/
+static AieRC XAie_AmdairIO_Write32(void *IOInst, u64 RegOff, u32 Value)
+{
+	char buf[20];
+	XAie_AmdairIO *AmdairIOInst = (XAie_AmdairIO *)IOInst;
+
+	XAIE_DBG("W: 0x%lx, 0x%x\n", RegOff, Value);
+
+	int address = open(AmdairIOInst->AmdairAddress, O_WRONLY);
+	if (address == -1) {
+		XAIE_ERROR("Error opening %s\n", AmdairIOInst->AmdairAddress);
+		return XAIE_ERR;
+	}
+	sprintf(buf, "0x%lx", RegOff);
+	write(address, buf, strlen(buf) + 1);
+	close(address);
+
+	int value = open(AmdairIOInst->AmdairValue, O_WRONLY);
+	if (value == -1) {
+		XAIE_ERROR("Error opening %s\n", AmdairIOInst->AmdairValue);
+		return XAIE_ERR;
+	}
+	sprintf(buf, "0x%x", Value);
+	write(value, buf, strlen(buf) + 1);
+	close(value);
+
+	return XAIE_OK;
+}
+
+/*****************************************************************************/
+/**
+*
+* This is the memory IO function to read 32bit data from the specified address.
+*
+* @param	IOInst: IO instance pointer
+* @param	RegOff: Register offset to read from.
+* @param	Data: Pointer to store the 32 bit value
+*
+* @return	XAIE_OK on success.
+*
+* @note		None.
+*
+*******************************************************************************/
+static AieRC XAie_AmdairIO_Read32(void *IOInst, u64 RegOff, u32 *Data)
+{
+	char buf[20];
+	XAie_AmdairIO *AmdairIOInst = (XAie_AmdairIO *)IOInst;
+
+	int address = open(AmdairIOInst->AmdairAddress, O_WRONLY);
+	if (address == -1) {
+		XAIE_ERROR("Error opening %s\n", AmdairIOInst->AmdairAddress);
+		return XAIE_ERR;
+	}
+	sprintf(buf, "0x%lx", RegOff);
+	write(address, buf, strlen(buf) + 1);
+	close(address);
+
+	int value = open(AmdairIOInst->AmdairValue, O_RDONLY);
+	if (value == -1) {
+		XAIE_ERROR("Error opening %s\n", AmdairIOInst->AmdairValue);
+		return XAIE_ERR;
+	}
+	read(value, buf, strlen(buf) + 1);
+	close(value);
+	*Data = strtoul(buf, 0, 0);
+
+	XAIE_DBG("R: 0x%lx = 0x%x\n", RegOff, *Data);
+
+	return XAIE_OK;
+}
+
+/*****************************************************************************/
+/**
+*
+* Write a value to a register by first clearing some bits from the register.
+* The bits are cleared by reading the current value and applying the mask.
+* Then the value is updated and written back to the register.
+*
+* @param	IOInst: IO instance pointer
+* @param	RegOff: Offset of the register to update.
+* @param	Mask: Mask to be applied to Data.
+* @param	Value: 32-bit data to be written.
+*
+* @return	None.
+*
+* @note		None.
+*
+*******************************************************************************/
+static AieRC XAie_AmdairIO_MaskWrite32(void *IOInst, u64 RegOff, u32 Mask,
+		u32 Value)
+{
+	u32 RegVal;
+
+	AieRC rc = XAie_AmdairIO_Read32(IOInst, RegOff, &RegVal);
+	if (rc != XAIE_OK)
+		return rc;
```

Comment:
nit: `XAIE_ERROR` here.

Reply:
Addressed in [6de16c1](https://github.com/stephenneuendorffer/aie-rt/pull/4/commits/6de16c1da98d0def29a457547e30e8c1a6f0bb43)